Bump pyo3 from 0.20.0 to 0.22.2

weiji14 opened the pull request:
Bumps [pyo3](https://github.com/pyo3/pyo3) from 0.20.0 to 0.22.2.
- [Release notes](https://github.com/pyo3/pyo3/releases)
- [Changelog](https://github.com/PyO3/pyo3/blob/main/CHANGELOG.md)
- [Commits](https://github.com/pyo3/pyo3/compare/v0.20.0...v0.22.2)

References:
- Migration guide at https://pyo3.rs/v0.21.1/migration#from-020-to-021 to use new `Bound<'py, T>` smart pointer
- Migration guide at https://pyo3.rs/v0.22.2/migration#pyclone-is-now-gated-behind-the-py-clone-feature to use the `py-clone` feature flag
- https://polar.sh/davidhewitt/posts/replacing-pyo3-api-pt1

weiji14 committed:
Bump pyo3 from 0.20.0 to 0.21.2

Bumps [pyo3](https://github.com/pyo3/pyo3) from 0.20.3 to 0.21.2.
- [Release notes](https://github.com/pyo3/pyo3/releases)
- [Changelog](https://github.com/PyO3/pyo3/blob/main/CHANGELOG.md)
- [Commits](https://github.com/pyo3/pyo3/compare/v0.20.3...v0.21.2)

weiji14 committed:
Bump pyo3 from 0.21.2 to 0.22.2 with py-clone feature

Bumps [pyo3](https://github.com/pyo3/pyo3) from 0.21.2 to 0.22.2.
- [Release notes](https://github.com/pyo3/pyo3/releases)
- [Changelog](https://github.com/PyO3/pyo3/blob/main/CHANGELOG.md)
- [Commits](https://github.com/pyo3/pyo3/compare/v0.21.2...v0.22.2)

weiji14 commented on src/lib.rs:
```diff
@@ -56,15 +57,16 @@ struct LazVlr {
     #[classmethod]
-    fn new_for_compression(
-        _cls: &PyType,
+    #[pyo3(signature = (point_format_id, num_extra_bytes, use_variable_size_chunks=false))]
```
Assuming that `use_variable_size_chunks=False` is the default, given the `use_variable_size_chunks.unwrap_or(false)` call at L78 below.

tmontaigu replied:
Doesn't that mean we could remove the option ?

weiji14 replied:
If it's just false/true, we could make it a `bool` instead of `Option`?

tmontaigu replied:
yes

weiji14 replied:
Ok, changed to `bool` at commit 56753f78f054bacb12724f7ef2102084a59c18fc.

weiji14 commented on Cargo.toml:
```diff
@@ -16,5 +16,5 @@ version = "0.9.2"
-version = "0.20.0"
-features = ["extension-module"]
+version = "0.22.2"
+features = ["extension-module", "py-clone"]
```
Setting the `py-clone` feature flag here (following https://pyo3.rs/v0.22.2/migration#pyclone-is-now-gated-behind-the-py-clone-feature) to avoid these errors in pyo3 0.22:

```rust
error[E0277]: the trait bound `Py<PyAny>: Clone` is not satisfied
  --> src/adapters.rs:35:5
   |
33 | #[derive(Clone)]
   |          ----- in this derive macro expansion
34 | pub(crate) struct PyFileObject {
35 |     file_obj: pyo3::PyObject,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `Py<PyAny>`
   |
   = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Py<PyAny>: Clone` is not satisfied
  --> src/adapters.rs:36:5
   |
33 | #[derive(Clone)]
   |          ----- in this derive macro expansion
...
36 |     write_fn: Option<pyo3::PyObject>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `Py<PyAny>`, which is required by `Option<Py<PyAny>>: Clone`
   |
   = note: required for `Option<Py<PyAny>>` to implement `Clone`
   = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Py<PyAny>: Clone` is not satisfied
  --> src/adapters.rs:37:5
   |
33 | #[derive(Clone)]
   |          ----- in this derive macro expansion
...
37 |     read_fn: Option<pyo3::PyObject>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `Py<PyAny>`, which is required by `Option<Py<PyAny>>: Clone`
   |
   = note: required for `Option<Py<PyAny>>` to implement `Clone`
   = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Py<PyAny>: Clone` is not satisfied
  --> src/adapters.rs:38:5
   |
33 | #[derive(Clone)]
   |          ----- in this derive macro expansion
...
38 |     readinto_fn: Option<pyo3::PyObject>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `Py<PyAny>`, which is required by `Option<Py<PyAny>>: Clone`
   |
   = note: required for `Option<Py<PyAny>>` to implement `Clone`
   = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `lazrs` (lib) due to 4 previous errors
```

tmontaigu replied:
Seems like it should be fine as the only time this clone is needed we hold the GIL

tmontaigu reviewed (commented):
seems like I forgot to hit submit

weiji14 committed:
Make use_variable_size_chunks param a bool type instead of Option<bool>

Default value for use_variable_size_chunks is now False instead of None, and so the `.unwrap_or(false)` call can be removed.

tmontaigu committed:
Bump upload artifact to v4

tmontaigu commented:
Thanks for this

tmontaigu commented:
Should I wait for the official release of Python3.13 or is a new release needed now ?

weiji14 commented:
Up to you. Maybe you can add Python 3.13 to CI first and then do a release?
